File: packages/contracts/python/noa_contracts/canonical_building_model.py
from __future__ import annotations

from typing import Literal, Any
from pydantic import BaseModel

from noa_contracts.common import ObservationType


class Level(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class Wall(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class Slab(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class RoofElement(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class Door(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class Window(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class Stair(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class Column(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class Beam(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class Room(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class MeshRef(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class Primitive(BaseModel):
    id: str
    model_config = {"extra": "allow"}


class ElementProvenance(BaseModel):
    element_id: str
    model_config = {"extra": "allow"}


class Elements(BaseModel):
    levels: list[Level]
    walls: list[Wall]
    slabs: list[Slab]
    roofs: list[RoofElement]
    doors: list[Door]
    windows: list[Window]
    stairs: list[Stair]
    columns: list[Column]
    beams: list[Beam]
    rooms: list[Room]


class GeometryLayer(BaseModel):
    meshes: list[MeshRef]
    parametric_primitives: list[Primitive]


class CanonicalBuildingModel(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    id: str
    project_id: str
    spatial_graph_id: str
    style_profile_id: str | None
    elements: Elements
    geometry_layer: GeometryLayer
    provenance: list[ElementProvenance]
    observation_types: dict[str, ObservationType]
    version: int
    created_at: str
    updated_at: str

File: packages/contracts/v1/canonical-building-model.ts
import type { UUID, ISO8601 } from "./common.js";
import type { ObservationType } from "./common.js";

export type ElementId = string;

export interface Level {
  id: string;
  [key: string]: unknown;
}

export interface Wall {
  id: string;
  [key: string]: unknown;
}

export interface Slab {
  id: string;
  [key: string]: unknown;
}

export interface RoofElement {
  id: string;
  [key: string]: unknown;
}

export interface Door {
  id: string;
  [key: string]: unknown;
}

export interface Window {
  id: string;
  [key: string]: unknown;
}

export interface Stair {
  id: string;
  [key: string]: unknown;
}

export interface Column {
  id: string;
  [key: string]: unknown;
}

export interface Beam {
  id: string;
  [key: string]: unknown;
}

export interface Room {
  id: string;
  [key: string]: unknown;
}

export interface MeshRef {
  id: string;
  [key: string]: unknown;
}

export interface Primitive {
  id: string;
  [key: string]: unknown;
}

export interface ElementProvenance {
  element_id: string;
  [key: string]: unknown;
}

export interface CanonicalBuildingModel {
  schema_version: "1.0";
  id: UUID;
  project_id: UUID;
  spatial_graph_id: UUID;
  style_profile_id: UUID | null;
  elements: {
    levels: Level[];
    walls: Wall[];
    slabs: Slab[];
    roofs: RoofElement[];
    doors: Door[];
    windows: Window[];
    stairs: Stair[];
    columns: Column[];
    beams: Beam[];
    rooms: Room[];
  };
  geometry_layer: {
    meshes: MeshRef[];
    parametric_primitives: Primitive[];
  };
  provenance: ElementProvenance[];
  observation_types: Record<ElementId, ObservationType>;
  version: number;
  created_at: ISO8601;
  updated_at: ISO8601;
}
